### PR TITLE
Rewrites the ts-npm-lint task such that we no longer use the external…

### PR DESCRIPTION
**Pull request by nickpape-msft**
* Since we were only using a small bit of ts-npm-lint, I first factored that code out & removed the dependency.
* Secondly, remodeled this code as a gulp.src/gulp.dest instead of glob/fs
* Thirdly, I updated the RegEx to be more lenient
* Fourthly, I added a cache so only files which have changed will get relinted

### CODE REVIEW

File: src/TSNpmLintTask.ts [hunk 1]
```diff
@@ -17,6 +21,69 @@ export class TSNpmLintTask extends GulpTask<ITSNPMLintTaskConfig> {
...
+    const filePattern: string = path.join(taskScope.buildConfig.libFolder, '**', '*.d.ts');
+
+    /**
+     * Matches:
+     *  /// <reference path="../../typings.d.ts" />
+     *  /// <reference path='../../typings.d.ts' />
+     *  ///<reference path='../../typings.d.ts' />
+     *  ///<reference path='../../typings.d.ts'/>
+     */
+    const referencePathRegex: RegExp = /^\/\/\/[ ]?<reference path=['"](.*)['"][ ]?\/>/gm;
```
Comment (iclanton) [inline src/TSNpmLintTask.ts]: Can probably be: ````/^\/\/\/[ ]?<reference[ ]+path=['"](.*)['"][ ]?\/>/gm;````

File: src/TSNpmLintTask.ts [hunk 2]
```diff
@@ -17,6 +21,69 @@ export class TSNpmLintTask extends GulpTask<ITSNPMLintTaskConfig> {
...
+          try {
+            const rawContents: string = file.contents.toString(encoding);
+            const relativePathToCurrentFile: string = path.relative(taskScope.buildConfig.rootPath, file.path);
+            taskScope.logVerbose(relativePathToCurrentFile);
+
+            /* tslint:disable:no-string-literal */
+            file[taskScope.name] = {
+              failureCount: 0
+            };
+            /* tslint:enable:no-string-literal */
```
Comment (iclanton) [inline src/TSNpmLintTask.ts]: This isn't a string literal, is it?

File: src/TSNpmLintTask.ts [hunk 3]
```diff
@@ -17,6 +21,69 @@ export class TSNpmLintTask extends GulpTask<ITSNPMLintTaskConfig> {
   };
 
   public executeTask(gulp: gulp.Gulp): void {
-    tsdLinter.fixTypings();
+    const taskScope: TSNpmLintTask = this;
+
+    const filePattern: string = path.join(taskScope.buildConfig.libFolder, '**', '*.d.ts');
```
Comment (iclanton) [inline src/TSNpmLintTask.ts]: Don't you want this to run over *.ts files too?

**Comment by iclanton** [comment]
Will the performance of the task in https://github.com/dzearing/gulp-core-build-typescript/blob/master/src/TSLintTask.ts get a benefit from this? I realize I'm actually not sure what these two tasks are for (and why this one is necessary, and when it gets called). Mind giving a brief explanation?

File: src/TSNpmLintTask.ts (hunk 3 above)
Comment (nickpape-msft) [inline src/TSNpmLintTask.ts]: I just replicated the functionality from ts-npm-lint, which only runs on the d.ts files.

File: src/TSNpmLintTask.ts (hunk 1 above)
Comment (nickpape-msft) [inline src/TSNpmLintTask.ts]: What about 

`^\/\/\/[ ]+<reference.*path=['"]([^'"]*)['"][^>]+>`

**Comment by nickpape-msft** [comment]
@iclanton No I checked in updates to the regular lint task the other day (in fact I have PR #16, which modifies some of that code, currently out) which introduce caching there. This just strips /// <reference it for the .d.ts files which are created as a part of the build result. TsNpmLint task might even be a bad name since we are no longer using that dependency.. Perhaps `StripTripleSlashReference` or something?

File: src/TSNpmLintTask.ts (hunk 2 above)
Comment (nickpape-msft) [inline src/TSNpmLintTask.ts]: I had it as a string literal but decided to use the taskname instead. Removing linter override.

File: src/TSNpmLintTask.ts (hunk 1 above)
Comment (iclanton) [inline src/TSNpmLintTask.ts]: Even better.

Comment (iclanton) [inline src/TSNpmLintTask.ts]: Even better.

**Comment by iclanton** [comment]
That makes sense, and I like the `StripTripleSlashReference` idea. Can you change it to that?

**Comment by iclanton** [comment]
Approved - but can you change the name?

**Pull request merged**